### ISSUE TITLE
Type checking in Firebase cloud functions

### ISSUE THREAD

**Issue by amyjko** [wordplaydev/wordplay#629]
### What are you trying to do that you can't?

We define many callable Firebase Cloud functions in `functions/src/index.ts`, but the types are often not declared and types are not shared between the client and the cloud functions. Instead, we're currently defining them in `/src/db/functions.ts` and `/functions/src/types.ts` and copying them manually. Yuck.

### What is your idea?

There are ways of defining shared libraries and allowing them to be used on both client and server. Explore these possibilities, define the correct types for all functions, fix any defects we find, and document how it works.

**Comment by annieli555** [comment]
Hi @amyjko! I'm interested in working on this issue. May I be assigned?

**Comment by amyjko** [comment]
Assigned! This one requires some research on best practices and then an implementation. I found a few online, but they were old, so I'm not sure if there are better solutions.

**Comment by annieli555** [comment]
Hi Amy!

Apologies for the delay here. I investigated sharing types between cloud functions and the client, and found a few common solutions mentioned. I've opened [a PR](https://github.com/wordplaydev/wordplay/pull/694) implementing option 1 because it felt the cleanest, but am listing them all here for completeness.

1. **Set up a local npm package**. This approach was suggested by this [answer in Reddit](https://www.reddit.com/r/Firebase/comments/xozd0s/comment/iq4fi7t/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button). It sounded like others in the thread had trouble getting this to work in their environments, but I was able to get the types to resolve locally.

2. **Copy the file during build time** (settled on by OP in the same [thread above](https://www.reddit.com/r/Firebase/comments/xozd0s/does_anyone_have_an_example_of_using_shared_types/)). This feels hacky to me, but seems like an option that we could do as a last resort.

3. **Use a symlink**. Unfortunately, I wasn't able to get this to work (I might have misconfigured something - I tried linking wordplay src from functions and wasn't able to get the reference to src/db/functions to resolve). Also, according to this [Stack Overflow post](https://stackoverflow.com/questions/66239777/deploying-firebase-functions-with-local-depencies-using-firebase-cli), they tried something similar under "Approach 1" but got errors trying to deploy functions.

4. **Firelink**. I saw this mentioned a few times, but it seems like this approach is not feasible if functions have dev dependencies, which ours does (see [this GitHub comment](https://github.com/vercel/turborepo/issues/1121#issuecomment-1327716054)). So I didn't pursue this approach.

**Comment by amyjko** [comment]
@annieli555 Thanks! I look forward to reviewing the PR soon.
